[AIRFLOW-2534] Fix bug in HiveServer2Hook

gglanzani opened the pull request:
(I'll be waiting on Travis to finish successfully before commenting here)

### JIRA
- [x] My PR addresses the following [Airflow JIRA](https://issues.apache.org/jira/browse/AIRFLOW-2534) issues and references them in the PR title.

### Description
- [x] Here are some details about my PR, including screenshots of any UI changes:

`HiveServer2Hook.get_results` is overwriting records when multiple sql statements are passed. This might be superflous when the first statements are just `SET`, but in principle we want to check if that's also the case. This PR addresses that, plus it refactors some existing Python code.

### Tests
- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

This PR modifies existing methods. The current test suite should catch regressions.

### Commits
- [x] My commits all reference JIRA issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
    1. Subject is separated from body by a blank line
    2. Subject is limited to 50 characters
    3. Subject does not end with a period
    4. Subject uses the imperative mood ("add", not "adding")
    5. Body wraps at 72 characters
    6. Body explains "what" and "why", not "how"


### Documentation
- [x] In case of new functionality, my PR adds documentation that describes how to use it.
    - No new functionality added.


### Code Quality
- [x] Passes `git diff upstream/master -u -- "*.py" | flake8 --diff`

KevinYang21 commented on airflow/hooks/hive_hooks.py:
```diff
@@ -740,6 +739,7 @@ class HiveServer2Hook(BaseHook):
     """
     def __init__(self, hiveserver2_conn_id='hiveserver2_default'):
         self.hiveserver2_conn_id = hiveserver2_conn_id
+        self._cursor_description = []
```
IMO we don't need to expose the cursor concept at the class level--it can also be not thread safe if we use and reset ourselves (this part I need to read a bit more on python myself to be sure). Maybe just return a tuple in _get_results()?

gglanzani replied:
Hi @yrqls21, thanks for the review!

I know about thread-safety alas, but after discussing with @Fokko, his idea was that Airflow never reuse the same class instance over multiple threads. 

@bolkedebruin or someone else: can you confirm?

As for returning a tuple, it had the issue of handling `to_csv`. Right now `to_csv` can work in batches. To reuse the same code for `to_csv` and for `get_results`, I use 

```python
for row in cur:
    yield row
```

Since `cur.arraysize = fetch_size`, I am batching automatically.

If I'd return instead of yielding, I wouldn't be batching anymore (you can still do that by splitting the methods for `to_csv` and `get_results`.

An alternative is raising an exception at the end of `_get_results`. That way I could send data after the `yield` in a Python 2/3 compatible way (I can't return from generators in Python 2 AFAIK).

The caller would then have to catch the exception to get the cursor description.

If you feel very strongly about `self._cursor_description`, I can therefore try that route.

Fokko replied:
If the cursor can be changed to a local variable, that would be better of course.

Personally I don't see how multiple threads use the same hook, or this might be implemented by hand.

KevinYang21 replied:
I don't think there will be such case neither. The only thing I can think of is that people making such calls in their python callback. This raises the question that do we want people to use the hook directl (people in Airbnb is doing this a lot)?

KevinYang21 commented on airflow/hooks/hive_hooks.py:
```diff
@@ -766,32 +766,47 @@ def get_conn(self, schema=None):
+                        self.log.debug("get_results returned no records")
+                    description = cur.description
+                    previous_description = self._cursor_description
+                    # if we return data, we expect the schema to be equal
```
I'm not sure if users are assuming the same thing at the beginning. Maybe add put the expectation in pydoc?

gglanzani replied:
Well, I certainly hope they assume consistent results to come back, right?

Currently the code only enforces this if the statement starts with `with` or `select`. I think they are the only statement where you'd be interesting in getting the results.

But adding something in the pydoc is certainly an option! Should I put it in every function calling `_get_results`?

KevinYang21 replied:
Ya that would be great (I'm one of those who heavily rely on pydoc before using any lib func).

KevinYang21 reviewed (commented):
Thank you for spotting this, we're also trying to increase the usage on this hook. +1 for writing all the tests--I failed to do that because of the package conflict (curious how you solved that).

KevinYang21 commented on airflow/hooks/hive_hooks.py:
```diff
@@ -802,29 +818,35 @@ def to_csv(
+            raise ValueError(message)
+
+        self.log.info("Done. Loaded a total of %s rows.", i)
+        self._cursor_description = []
```
some legacy? :P

Fokko reviewed (changes requested):
Also we need to fix the tests on PyHive since it is currently not compatible with SQLAlchemy 1.2: https://github.com/Fokko/PyHive/pull/1

We need this because this is also the version that Airflow is on. I've been working on this together with @gglanzani But the SQLAlchemy dialect part is hard to grasp since there aren't a lot of examples how this is done. 

It is still failing on one test: https://travis-ci.org/Fokko/PyHive/jobs/388325847#L2277

This is because the date isn't parsed correctly. `'2018-01-01 00:00:00'` instead of `datetime.datetime(2018, 1, 1, 0, 0, 0)`.

I've dropped a question on the SQLAlchemy discussions https://groups.google.com/forum/#!topic/sqlalchemy/evjPjtbL_Pw. Turns out the functionality that we where relying on has been deprecated. I'm also fine with leaving the strings as is.

Fokko commented on setup.py:
```diff
@@ -156,7 +156,7 @@ def write_version(filename=os.path.join(*['airflow',
     'pyhive>=0.1.3',
     'impyla>=0.13.3',
-    'thrift_sasl==0.2.1',
+    'thrift_sasl==0.2.1'
```
I strongly believe we should move to `thrift_sasl` version `0.3.0`. This is the one that they are using as well: https://github.com/dropbox/PyHive/blob/master/dev_requirements.txt#L17

Fokko replied:
Could you remove this dependency? This is pulled in by the PyHive dependency recursively.

gglanzani committed:
[AIRFLOW-2534] Fix bug in HiveServer2Hook

This commit also adds numerous tests for HiveServer2 and switches
Impyla for PyHive (0.6.0), making HiveServer2 Python 2 compatible.

Fokko commented on setup.py:
```diff
@@ -155,9 +155,9 @@ def write_version(filename=os.path.join(*['airflow',
+    'pyhive>=0.6.0',
     'impyla>=0.13.3',
-    'thrift_sasl==0.2.1',
+    'thrift_sasl>=0.3.0'
```
@gglanzani This dependency is transitive, we can just remove it from here.

Fokko reviewed (approved):
Awesome! Merging!
